eq_cir_converter_service/utils/helper_utils.py

```diff
@@ -0,0 +1,181 @@
+REGEX_PARA_SPLIT = re.compile(r"<p>(.*?)</p>", flags=re.IGNORECASE | re.DOTALL)
+REGEX_PLACEHOLDER = re.compile(r"\{(.*?)\}", flags=re.IGNORECASE)
+
+
+# --- HTML Tag Processing ---
+def clean_html_tags(text: str) -> str:
+    """Cleans HTML tags from the text, replacing <b> with <strong> and removing <br> and <p> tags.
+
+    :param text: The input text containing HTML tags.
+    :return: The cleaned text with HTML tags removed or replaced.
+    """
+    text = REGEX_B_OPEN.sub("<strong>", text)  # Replace <b> with <strong>
+    text = REGEX_B_CLOSE.sub("</strong>", text)  # Replace </b> with </strong>
+    text = REGEX_BR_TAGS.sub("", text)  # Remove <br> tags
+    text = REGEX_P_TAGS.sub("", text)  # Remove <p> tags
+    return text.strip()
+
+
+# --- Paragraph Extraction ---
+def extract_paragraphs(html: str) -> list[str]:
+    """Extracts paragraphs from HTML content, returning a list of cleaned paragraph strings.
+
+    :param html: The HTML content to extract paragraphs from.
+    :return: A list of cleaned paragraphs.
+    """
+    return REGEX_PARA_SPLIT.findall(html)
+
+
+# --- Placeholder Extraction ---
+def extract_placeholders(text: str) -> list[str]:
+    """Extracts placeholders from the text, returning a list of placeholder names.
+
+    :param text: The input text containing placeholders.
+    :return: A list of placeholder names.
+    """
+    return REGEX_PLACEHOLDER.findall(text)
+
+
+# --- Handle "text" with "placeholders" ---
+def split_text_with_placeholders(
+    text_obj: dict[str, str | list | object],
+) -> list[str | dict[str, str | list | object]]:
+    """Splits the 'text' field in the text_obj into paragraphs, cleaning HTML tags and extracting placeholders.
+
+    :param text_obj: A dictionary containing 'text' and 'placeholders'.
+    :return: A list of cleaned paragraphs or dictionaries with placeholders.
+    """
+    raw_text = text_obj.get("text", "")
+    placeholders = text_obj.get("placeholders", [])
+    paragraphs = extract_paragraphs(str(raw_text))
+
+    result: list[str | dict[str, str | list | object]] = []
+    for para in paragraphs:
+        cleaned = clean_html_tags(para).strip()
+        if not cleaned:
+            continue
+
+        used_counts = Counter(extract_placeholders(cleaned))
+        relevant: list[dict] = []
+        for ph_name, count in used_counts.items():
+            if isinstance(placeholders, list):
+                matching = []
+                for placeholder in placeholders:
+                    if isinstance(placeholder, dict) and placeholder.get("placeholder") == ph_name:
+                        # Ensure the placeholder is a dictionary with a 'placeholder' key
+                        matching = [placeholder]
+
+                # If the placeholder is found, create a deep copy for each occurrence
+                if len(matching) > 0:
+                    relevant.extend(deepcopy(matching[0]) for _ in range(count))
```

Comment:
Why not to create the deepcopy beforehand for readability? `matching = deepcopy(placeholder)`